haystack_experimental/components/tools/tool_invoker.py

```diff
@@ -0,0 +1,204 @@
+        """
+        Initialize the ToolInvoker component.
+
+        :param tools:
+            A list of tools that can be invoked.
+        :param raise_on_failure:
+            If True, the component will raise exceptions in case the tool is not found or invocation fails.
+            If False, the component will return a ChatMessage from the tool, wrapping the error message.
+        """
+        if not tools:
+            raise ValueError("ToolInvoker requires at least one tool to be provided.")
+
+        self.tools = tools
+        self._tools_with_names = {tool.name: tool for tool in tools}
+        self.raise_on_failure = raise_on_failure
+
+    @staticmethod
+    def _convert_tool_result_to_string(result: Any) -> str:
+        """
+        Converts the tool invocation result to a string.
+
+        :param result:
+            The tool result.
+        :returns:
+            The string representation of the tool result.
+        """
+        try:
+            return json.dumps(result)
+        except TypeError as e:
+            logger.warning(
+                "Failed to convert tool result to string using `json.dumps`. Error: {error}. Falling back to `str`.",
+                error=e,
+            )
+            return str(result)
+
+    @component.output_types(tool_messages=List[ChatMessage])
+    def run(self, message: ChatMessage):
+        """
+        Processes a ChatMessage containing tool calls and invokes the corresponding tools, if available.
+
+        :param message:
+            A ChatMessage from assistant, containing prepared tool calls.
+        :returns:
+            A dictionary with the key `tool_messages` containing a list of ChatMessage objects with tool role
+            that contain the results of the tool invocations.
+
+        :raises ValueError:
+            If the message is not from the assistant role or does not contain tool calls.
+        :raises ToolNotFoundException:
+            If the tool is not found in the list of available tools.
+        :raises ToolInvocationError:
+            If the tool invocation fails.
+        """
+        if not message.is_from(ChatRole.ASSISTANT):
+            raise ValueError("ToolInvoker only supports messages from the assistant role.")
+
+        tool_calls = message.tool_calls
+        if not tool_calls:
+            raise ValueError("ToolInvoker only supports messages with tool calls.")
+
+        tool_messages = []
+
+        for tool_call in tool_calls:
+            tool_name = tool_call.tool_name
+            tool_arguments = tool_call.arguments
+
+            if not tool_name in self._tools_with_names:
+                msg = _TOOL_NOT_FOUND.format(tool_name=tool_name, available_tools=self._tools_with_names.keys())
+                if self.raise_on_failure:
+                    raise ToolNotFoundException(msg)
+                tool_messages.append(ChatMessage.from_tool(tool_result=msg, origin=tool_call))
```

Comment:
I think it would be useful to have a flag in the tool call result payload to indicate errors.

Reply:
makes sense
done!